enumerable/queue.go

```diff
@@ -0,0 +1,113 @@
+package enumerable
+
+// Queue is an extention of the enumerable interface allowing individual
+// items to be added into the enumerable.
+//
+// Added items will be yielded in a FIFO order.  Items may be added after
+// enumeration has begun.
+type Queue[T any] interface {
+	Enumerable[T]
+	// Put adds an item to the queue.
+	Put(T) error
+	// Size returns the current length of the backing array.
+	//
+	// This may include empty space where yield items previously resided.
+	// Useful for testing and debugging.
+	Size() int
+}
+
+// For now, increasing the size one at a time is likely optimal
+// for the only useage of the queue type.  We may wish to change
+// this at somepoint however.
+const growthRate int = 1
+
+type queue[T any] struct {
+	values       []T
+	currentIndex int
+	lastSetIndex int
+	zeroIndexSet bool
+}
+
+var _ Queue[any] = (*queue[any])(nil)
+
+// NewQueue creates an empty FIFO queue.
+//
+// It is implemented using a dynamically sized ring-buffer.
+func NewQueue[T any]() Queue[T] {
+	return &queue[T]{
+		values:       []T{},
+		currentIndex: -1,
+		lastSetIndex: -1,
+	}
+}
+
+func (q *queue[T]) Put(value T) error {
+	var index int
+	if !q.zeroIndexSet {
+		// If the zero-index is empty, we should use it - circling
+		// the ring buffer back to the beginning.
+		index = 0
+		q.zeroIndexSet = true
+	} else {
+		index = q.lastSetIndex + 1
+	}
+
+	if index >= len(q.values) {
+		newValues := make([]T, len(q.values)+growthRate)
+		copy(newValues, q.values[:index])
+		q.values = newValues
+	} else if index == q.currentIndex+growthRate {
+		// If the write index has caught up to the read index
+		// the new value needs to be written between the two
+		// e.g: [3,4,here,1,2]
```

Comment:
question: why not rearrange the elements?

I see now the reason why you grow the size by 1: it's easier to handle this scenario.
I would say when this happens we can rearrange from:
```
[3,4,1,2]
     ^
     |
     read and write index
```
to this:
```
[1,2,3,4,_,_,_]
 ^       ^
 |       |
 |       write index
 read index
```
The queue is not thread-safe anyway

Reply:
That is an interesting thought. Will look at this some more. I do worry that this might make an future threadsafetiness harder/more-expensive to achieve, but maybe not really anymore than allowing the buffer to grow already does.

What benefit do you see in this though?

Reply:
I though this was the main reason why you chose (for now) to grow the array by 1. So that it's easier to manage indexes. Otherwise your read index will have to keep in mind that there are empty slots for future writing that it will have to jump over. 
```
[1,2,_,_,_,3,4]
   ^ ^     ^
   | |     next read index
   | write index
   read index
 ```
With this rearrangement you can grow it as much as you want.

Reply:
Note: I actually tried using this setup last week to help solve an edge case.  It did not simplify the code unfortunately.